File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/extension/RepositoryKoraExtension.java
package ru.tinkoff.kora.database.annotation.processor.extension;

import ru.tinkoff.kora.annotation.processor.common.AnnotationUtils;
import ru.tinkoff.kora.annotation.processor.common.NameUtils;
import ru.tinkoff.kora.database.annotation.processor.DbUtils;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;

import static ru.tinkoff.kora.annotation.processor.common.TagUtils.parseTagValue;
import static ru.tinkoff.kora.annotation.processor.common.TagUtils.tagsMatch;

public class RepositoryKoraExtension implements KoraExtension {
    private final Elements elements;
    private final Types types;

    public RepositoryKoraExtension(ProcessingEnvironment processingEnvironment) {
        this.elements = processingEnvironment.getElementUtils();
        this.types = processingEnvironment.getTypeUtils();
    }

    @Override
    public KoraExtensionDependencyGenerator getDependencyGenerator(RoundEnvironment roundEnvironment, TypeMirror typeMirror, String tag) {
        if (typeMirror.getKind() != TypeKind.DECLARED) {
            return null;
        }
        var element = this.types.asElement(typeMirror);
        if (element.getKind() != ElementKind.INTERFACE && (element.getKind() != ElementKind.CLASS || !element.getModifiers().contains(Modifier.ABSTRACT))) {
            return null;
        }
        if (AnnotationUtils.findAnnotation(element, DbUtils.REPOSITORY_ANNOTATION) == null) {
            return null;
        }
        if (!tagsMatch(tag, parseTagValue(element))) {
            return null;
        }
        var typeElement = (TypeElement) this.types.asElement(typeMirror);
        var repositoryName = NameUtils.generatedType(typeElement, "Impl");
        return KoraExtensionDependencyGenerator.generatedFromWithName(this.elements, typeElement, repositoryName);
    }
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/extension/RepositoryKoraExtension.kt
package ru.tinkoff.kora.database.symbol.processor.extension

import com.google.devtools.ksp.KspExperimental
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.ClassKind
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.google.devtools.ksp.symbol.Modifier
import ru.tinkoff.kora.database.symbol.processor.DbUtils
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionResult
import ru.tinkoff.kora.kora.app.ksp.extension.KoraExtension
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.TagUtils.parseTag
import ru.tinkoff.kora.ksp.common.TagUtils.tagMatches
import ru.tinkoff.kora.ksp.common.getOuterClassesAsPrefix

@KspExperimental
class RepositoryKoraExtension(private val kspLogger: KSPLogger) : KoraExtension {
    override fun getDependencyGenerator(resolver: Resolver, type: KSType, tag: String?): (() -> ExtensionResult)? {
        if (type.declaration !is KSClassDeclaration) {
            return null
        }
        val declaration = type.declaration as KSClassDeclaration
        if (declaration.classKind != ClassKind.INTERFACE && !(declaration.classKind == ClassKind.CLASS && Modifier.ABSTRACT in declaration.modifiers)) {
            return null
        }
        if (declaration.findAnnotation(DbUtils.repositoryAnnotation) == null) {
            return null
        }
        if (!tag.tagMatches(declaration.parseTag())) {
            return null
        }
        val repositoryName: String = declaration.getOuterClassesAsPrefix() + declaration.simpleName.asString() + "_Impl"
        return generatedByProcessorWithName(resolver, declaration, repositoryName)
    }
}
